ambry-api/src/main/java/com.github.ambry/network/BoundedNettyByteBufReceive.java

```diff
@@ -0,0 +1,129 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A netty {@link ByteBuf} version of Receive to buffer the incoming request or response.
+ */
+public class BoundedNettyByteBufReceive implements BoundedReceive<ByteBuf> {
+
+  private ByteBuf buffer = null;
+  private ByteBuf sizeBuffer = null;
+  private long sizeToRead = 0;
+  private long sizeRead = 0;
+  private Logger logger = LoggerFactory.getLogger(getClass());
+  private AtomicBoolean released = new AtomicBoolean(false);
+
+  public BoundedNettyByteBufReceive() {
+  }
+
+  @Override
+  public boolean isReadComplete() {
+    return !(buffer == null || sizeRead < sizeToRead);
+  }
+
+  /**
+   * Reading bytes from the {@link ReadableByteChannel} and writing those bytes to the given {@link ByteBuf}.
+   * This method would change the writerIndex of the given buffer.
+   * @param channel The {@link ReadableByteChannel} to read the bytes out.
+   * @param buffer The {@link ByteBuf} to write bytes to.
+   * @return The number of bytes read from channel.
+   * @throws IOException Any I/O error.
+   */
+  private int readBytesFromReadableByteChannel(ReadableByteChannel channel, ByteBuf buffer) throws IOException {
+    int n = channel.read(buffer.internalNioBuffer(buffer.writerIndex(), buffer.capacity() - buffer.writerIndex()));
+    buffer.writerIndex(buffer.writerIndex() + n);
+    return n;
+  }
+
+  @Override
+  public long readFrom(ReadableByteChannel channel) throws IOException {
+    long bytesRead = 0;
+    if (buffer == null) {
+      if (sizeBuffer == null) {
+        sizeBuffer = ByteBufAllocator.DEFAULT.heapBuffer(8);
+      }
+      try {
+        bytesRead = readBytesFromReadableByteChannel(channel, sizeBuffer);
+        if (bytesRead < 0) {
+          throw new EOFException();
+        }
+      } catch (IOException e) {
+        sizeBuffer.release();
+        throw e;
+      }
+      if (sizeBuffer.writerIndex() == sizeBuffer.capacity()) {
+        sizeToRead = sizeBuffer.readLong();
+        sizeRead += 8;
+        sizeBuffer.release();
+        buffer = ByteBufAllocator.DEFAULT.heapBuffer((int) sizeToRead - 8);
+      }
+    }
+    if (buffer != null && sizeRead < sizeToRead) {
+      long bytesReadFromChannel;
+      try {
+        bytesReadFromChannel = readBytesFromReadableByteChannel(channel, buffer);
+        if (bytesReadFromChannel < 0) {
+          throw new EOFException();
+        }
+      } catch (IOException e) {
+        buffer.release();
+        throw e;
+      }
+      sizeRead += bytesReadFromChannel;
+      bytesRead += bytesReadFromChannel;
+    }
+    logger.trace("size read from channel {}", sizeRead);
+    return bytesRead;
+  }
+
+  /**
+   * Returns the payload as {@link ByteBuf}, at the same time release the current reference to this payload.
+   * It's not safe to call this function again.
+   * @return
+   */
+  @Override
+  public ByteBuf getAndRelease() {
```

Comment:
Should this method return null if the buffer was not yet created or the buffer was released because of an error? Right now, it may throw an exception upon calling `retainedDuplicate()`